Src/IronPython.Modules/_codecs.cs

```diff
@@ -406,63 +406,109 @@ public static PythonTuple utf_7_encode(CodeContext context, string input, string
 
         #region Utf-32 Functions
 
-        public static PythonTuple utf_32_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors = "strict", bool final = false)
-            => DoDecode(context, "utf-32", Encoding.UTF32, input, errors, final).ToPythonTuple();
+        public static PythonTuple utf_32_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors = "strict", bool final = false) {
+            PythonTuple res = utf_32_ex_decode(context, input, errors, 0, final);
+            return PythonTuple.MakeTuple(res[0], res[1]);
+        }
 
         public static PythonTuple utf_32_encode(CodeContext context, string input, string errors = "strict")
-            => DoEncode(context, "utf-32", Encoding.UTF32, input, errors, true).ToPythonTuple();
-
-        #endregion
-
-        public static PythonTuple utf_32_ex_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors = "strict")
-            => utf_32_ex_decode(context, input, errors, null, null);
+            => DoEncode(context, "utf-32", Utf32LeBomEncoding, input, errors, includePreamble: true).ToPythonTuple();
+
+        public static PythonTuple utf_32_ex_decode(CodeContext context, [BytesConversion]IList<byte> input, string errors, int byteorder = 0, bool final = false) {
+
+            object output;
+            object numBytes;
+
+            if (byteorder != 0) {
```

Comment:
Seems like essentially the same thing in both cases. Maybe simplifying to:
```C#
if (byteorder == 0) byteorder = Utf32DetectByteorder(input);
var res = (byteorder > 0) ? ...
```

Reply:
It is similar, but not the same. When given `byteorder` is not 0, BOM (if present) is passed on to output. If `byteorder` is 0, BOM (if present) is skipped. I have added some tests in `Tests/modules/io_related/test_codecs.py` to catch it and verified the stest correctnets with running under CPython 3.4.

But I can replace calls to `utf_32_[lb]e_decode` by `DoDecode`. Apart from making the code more symmetrical (thus neater), it will make the return value of the call of the same type, hence simplifying unpacking of the tuple.

Reply:
Ah right, missed the BOM Encoding bit. The difference is clearer with both that use `DoDecode`. 👍